src/main/java/org/jenkinsci/plugins/vSphereCloudSlaveTemplate.java

```diff
@@ -349,6 +352,38 @@ protected Object readResolve() {
                 LOGGER.log(Level.CONFIG, " - Failed to reconfigure strategy", ex);
             }
         }
+        // Earlier versions of the code may have stored configuration in a
+        // different way but in the same kind of fields, so we need an explicit
+        // versioning to know how to mutate the data.
+        if( configVersion<=0) {
+            LOGGER.log(Level.CONFIG,
+                    "{0} loaded old configuration that had hard-coded underscore at the end of the cloneNamePrefix.",
+                    this);
+            // In version one, the underscore was removed from the code so we
+            // have to put it in the data (the user is then free to change it to
+            // something else if they want to).
+            this.cloneNamePrefix = this.cloneNamePrefix + "_";
```

Comment:
Can `this.cloneNamePrefix` be empty going into this method, e.g. perhaps if there was no config object at all to begin with? Is `readResolve()` only called if there is an existing config, and if so, is `cloneNamePrefix` guaranteed to be non-null and non-empty?

Reply:
There are no guarantees.
 - If it was empty then the slave node names would previously have been named "_someRandomId" and this code will ensure that this remains the case.
 - If it was null then the slave node names would previously have been named "null_someRandomId" and this code will ensure that this remains the case.
i.e. The old behavior is maintained, but a bad config won't be magically made good.

In an ideal world, we'd have some sort of "post deserialization fix-up" at the cloud level that ensured that, within the vSphereCloud, the prefix was non-null, non-empty and unique for each template, but I'm not 100% sure where that code would have to go / get-called-from in order to make it work in all situations.